### ISSUE TITLE
Missing GeoPtProperty?

### ISSUE THREAD

**Issue by dominicglenn** [potatolondon/djangae#955]
Hey,

I'm curious as to why you don't have a usable GeoPtProperty since it's functionality already included in NDB?

**Comment by adamalton** [comment]
I'm sure that when I looked at this a few years ago, geo stuff was only available in the Search API.  But looking at it again now, the Datastore API does provide a [GeoPt property type](https://cloud.google.com/appengine/docs/standard/python/refdocs/google.appengine.api.datastore_types#google.appengine.api.datastore_types.GeoPt).

So we could implement a field for that.

In terms of how that field could actually be used (i.e. how it could be queried), it would be nice if we could implement something along the lines of the [Geo Django module](https://docs.djangoproject.com/en/1.11/ref/contrib/gis/model-api/).  Or perhaps actually implement the database backend functionality to make that Geo Django stuff work.  But I don't think that the Datastore has the querying capability to do that.

Looking at the [NDB geo querying capabilities](https://cloud.google.com/appengine/docs/standard/python/search/query_strings#Python_Queries_on_geopoint_fields) it looks like the only thing it can do it query for points which are within a given distance from a given point.  So I'm thinking that a Djangae implementation would be something like:

* A custom `GeoPoint` model field, which stores a `datastore.GeoPt` property under the hood.
* A custom query operator (or whatever it's called), such as `__distance`, which would allow queries such as `.filter(my_geo_field__distance__lte=(10, (54.9273, -0.14252)))` where the first value is the distance and the second value is the point to measure from.
* Not implement anything else, because we probably can't support anything else on the Datastore.

Does this sound like what you are expecting?

(I think we could probably find a better syntax for the querying, but you get the idea.)

**Comment by seawolf42** [comment]
I wasn't the original asker, but that seems extremely reasonable and useful to me.

jeff

**Comment by dominicglenn** [comment]
@adamalton Yeah it sounds like what I was expecting. I can't find specific documentation on the different ways that you can query a GeoPt in the NDB documentation, is distance the only one we can do?

**Comment by adamalton** [comment]
Oh, woops.  The link that I put in my previous comment in for the **Search API** geo querying capabilities.  It seems that in the Datastore, you can store a geo point, but there isn't actually any querying functionality!  https://stackoverflow.com/questions/1033240/how-do-i-query-for-entities-that-are-nearby-with-the-geopt-property-in-google

That explains why I had looked at this and decided that we shouldn't bother!

However, there's a project called [geomodel](https://code.google.com/archive/p/geomodel/) which implements simple geo queries (bounding box, and distance).  So I think we could take a similar approach to that (or copy/lift the basic logic), and implement the same thing that I suggested previously.  It's going to be a bit of work though, which I don't really have time to do at the moment.  But if someone wants to take a stab at it, then I can try to provide guidance and rubber ducking.

As far as possible, we should probably make our querying syntax the same as this: https://docs.djangoproject.com/en/1.11/ref/contrib/gis/db-api/#distance-queries
